File: packages/providers/src/implementations/fallback.ts
import { LLMProvider, GenerateRequest, GenerateResponse, LLMProviderCapability } from '@gh-ai-workflows/core';

export class FallbackProvider implements LLMProvider {
  readonly providerId = 'fallback';
  readonly capabilities: LLMProviderCapability;

  constructor(
    private providers: LLMProvider[],
    options: {
      maxRetriesPerProvider?: number;
    } = {}
  ) {
    // Capabilities are the intersection of all providers, or the most capable one depending on strategy.
    // Here we take the capabilities of the first provider as the primary.
    this.capabilities = providers[0]?.capabilities || {
      capabilities: new Set(),
      maxTokens: 0,
      contextWindow: 0,
    };
  }

  async generate(request: GenerateRequest): Promise<GenerateResponse> {
    const errors: Error[] = [];

    for (const provider of this.providers) {
      try {
        return await provider.generate(request);
      } catch (e) {
        errors.push(e instanceof Error ? e : new Error(String(e)));
      }
    }

    throw new Error(`All providers failed. Errors: ${errors.map(e => e.message).join('; ')}`);
  }
}

File: crates/wfs-core/src/llm/fallback.rs
use async_trait::async_trait;
use std::collections::HashSet;

use crate::error::LlmError;
use crate::types::{GenerateRequest, GenerateResponse, ProviderCapabilities};
use super::LlmProvider;

pub struct FallbackProvider {
    providers: Vec<Box<dyn LlmProvider>>,
    caps: ProviderCapabilities,
}

impl FallbackProvider {
    pub fn new(providers: Vec<Box<dyn LlmProvider>>) -> Self {
        let caps = providers
            .first()
            .map(|p| p.capabilities().clone())
            .unwrap_or(ProviderCapabilities {
                capabilities: HashSet::new(),
                max_tokens: 0,
                context_window: 0,
            });
        Self { providers, caps }
    }
}

#[async_trait]
impl LlmProvider for FallbackProvider {
    fn provider_id(&self) -> &'static str {
        "fallback"
    }

    fn capabilities(&self) -> &ProviderCapabilities {
        &self.caps
    }

    async fn generate(&self, request: &GenerateRequest) -> Result<GenerateResponse, LlmError> {
        let mut errors = Vec::new();

        for provider in &self.providers {
            match provider.generate(request).await {
                Ok(response) => return Ok(response),
                Err(e) => {
                    errors.push(e.to_string());
                }
            }
        }

        Err(LlmError::Provider(format!(
            "All providers failed: {}",
            errors.join("; ")
        )))
    }
}
